File: src/main/java/br/com/zumbolovsky/fateapp/service/TimeoutAspect.java
package br.com.zumbolovsky.fateapp.service;

import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.springframework.stereotype.Component;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

@Aspect
@Component
public class TimeoutAspect {

    @Around("@annotation(br.com.zumbolovsky.fateapp.service.Timeout) && @annotation(timeout)")
    public Object around(ProceedingJoinPoint pjp, Timeout timeout) {
        var executor = Executors.newSingleThreadExecutor();
        var future = executor.submit(() -> {
            try {
                return (String) pjp.proceed();
            } catch (Throwable e) {
                throw new RuntimeException(e);
            }
        });
        try {
            return future.get(timeout.value(), TimeUnit.MILLISECONDS);
        } catch (TimeoutException e) {
            future.cancel(true);
            throw new RuntimeException("Timeout!", e);
        } catch (ExecutionException | InterruptedException e) {
            throw new RuntimeException(e);
        } finally {
            executor.shutdownNow();
        }
    }
}

File: src/main/kotlin/br/com/zumbolovsky/fateapp/web/TimeoutAspect.kt
package br.com.zumbolovsky.fateapp.web

import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.springframework.stereotype.Component
import java.util.concurrent.Executors
import java.util.concurrent.Future
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException


@Aspect
@Component
class TimeoutAspect {

    @Around("@annotation(br.com.zumbolovsky.fateapp.web.Timeout) && @annotation(timeout)")
    fun around(pjp: ProceedingJoinPoint, timeout: Timeout): Any {
        val executor = Executors.newSingleThreadExecutor()
        val future: Future<Any> = executor.submit<Any> { pjp.proceed() as String }
        return try {
            future.get(timeout.value, TimeUnit.MILLISECONDS)
        } catch (e: TimeoutException) {
            future.cancel(true)
            throw RuntimeException("Timeout!", e)
        } finally {
            executor.shutdownNow()
        }
    }
}